### PR TITLE
refactor(forms): match getError and hasError to get method signature

### LINKED ISSUE #19734: Change the getError and hasError signature to have more flexibility
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[X ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
Currently the methods [AbstractControl#getError](https://angular.io/api/forms/AbstractControl#getError) and [AbstractControl#hasError](https://angular.io/api/forms/AbstractControl#hasError) accept a 2nd. parameter that let you specify the path of the control. Ex:

`formGroup.hasError('errorCode', ANYPATH);`

However, both of these methods have the following signature:

```
getError(errorCode: string, path?: string[]): any

 hasError(errorCode: string, path?: string[]): boolean
```

... and behind the scenes these methods just call the [AbstractControl#get](https://angular.io/api/forms/AbstractControl#get) method, which one accepts a `string` or an `array` of `number | string`.

Doesn't it make sense to change the signatures to the same of `get` method?

```
getError(errorCode: string, path?: string | (number | string)[]): any

 hasError(errorCode: string, path?: string | (number | string)[]): boolean
```

So, we can just call this method as usual in `get`.

## Expected behavior
Instead of `formGroup.hasError('required', ['x', 'y', 'z'])` we can have `formGroup.hasError('required', 'x.y.z')` for example (as we can call in `get` method).

## What is the motivation / use case for changing the behavior?
More flexibility.

### PR DESCRIPTION
**Pull request by Toxicable**
closes #19734

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Feature
```

## What is the current behavior?
`getError` and `hasError` mismatch the api of `get`

Issue Number: #19734

## What is the new behavior?
apis match

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by kara — COMMENTED** [review]
Looks good, but can we also get a test for `hasError`, since we are changing it here?

**Comment by Toxicable** [comment]
@kara done

**Review by kara — APPROVED** [review]
LGTM

**Comment by kara** [comment]
@IgorMinar API review?

**Review by IgorMinar — CHANGES_REQUESTED** [review]

File: packages/forms/src/model.ts [hunk 1]
```diff
@@ -537,7 +537,9 @@ export abstract class AbstractControl {
    *
    * If no path is given, it checks for the error on the present control.
    */
-  hasError(errorCode: string, path?: string[]): boolean { return !!this.getError(errorCode, path); }
+  hasError(errorCode: string, path?: Array<string|number>|string): boolean {
```
[the remark this answers is not in the document]
Comment (Toxicable) [reply packages/forms/src/model.ts]: Seams valid if you're traversing through a form array.
Ill add some tests.

File: packages/forms/src/model.ts [hunk 2]
```diff
@@ -537,7 +537,9 @@ export abstract class AbstractControl {
    *
    * If no path is given, it checks for the error on the present control.
```
[the remark this answers is not in the document]
Comment (Toxicable) [reply packages/forms/src/model.ts]: Wasn't sure what to change here so I added a note on how the path can be used.
Let me know if you want something else

**Comment by Toxicable** [comment]
Travis failure looks like a flake to me

File: packages/forms/src/model.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (kara) [reply packages/forms/src/model.ts]: @Toxicable I took a look again, but wasn't able to find the docs that you added. Is the commit up to date?

[the remark this answers is not in the document]
Comment (Toxicable) [reply packages/forms/src/model.ts]: @kara Uhhh, I guess not sorry, Ill double check this afternoon

[the remark this answers is not in the document]
Comment (Toxicable) [reply packages/forms/src/model.ts]: @kara Added some detail on what the path is meant to be and how to use it

[the remark this answers is not in the document]
Comment (kara) [reply packages/forms/src/model.ts]: Looks good. I'll let @IgorMinar take one more look though, while it's presubmitting.

**Comment by kara** [comment]
[presubmit](http://test/OCL:186625804:BASE:186625815:1519319672973:9db50399)

**Review by IgorMinar — APPROVED** [review]

**Comment by IgorMinar** [comment]
thanks

**Comment by kara** [comment]
Some Google test failures that need to be investigated

**Comment by rafaelss95** [comment]
@kara sry to ping you, but do you have any news?

**Comment by kara** [comment]
@Toxicable Would you mind rebasing this? Sorry it's been sitting so long. If you can rebase, we can take another look at the test failures and see if they're still relevant.

**Comment by Toxicable** [comment]
@kara rebased! 
`ci/circleci: test_aio_local ` looks like a flake

**Comment by kara** [comment]
[new presubmit](https://test.corp.google.com/ui#id=OCL:223420403:BASE:223420525:1543533441617:26cec8df)

File: packages/forms/src/model.ts [hunk 3]
```diff
@@ -664,7 +664,9 @@ export abstract class AbstractControl {
    * control.
    * @returns True when the control with the given path has the error, otherwise false.
    */
-  hasError(errorCode: string, path?: string[]): boolean { return !!this.getError(errorCode, path); }
+  hasError(errorCode: string, path?: Array<string|number>|string): boolean {
```
Comment (kara) [inline packages/forms/src/model.ts]: Can you update the type signature on `AbstractControlDirective` as well? Looks like it's inconsistent.

**Review by kara — CHANGES_REQUESTED** [review]

File: packages/forms/src/model.ts (hunk 3 above)
Comment (Toxicable) [reply packages/forms/src/model.ts]: oh yes, will get on that

**Comment by Toxicable** [comment]
@kara Hey, sorry the turn around from my side has been so slow.
updated those signatures aswell and re-rebased, let me know if I missed anything else

**Review by kara — APPROVED** [review]
LGTM. This last change fixed the internal failures (and don't worry about timing, you are still doing better than we are :) ).

**Comment by kara** [comment]
@IgorMinar I think it needs one more review from you since the public-API changed again.

**Review by IgorMinar — APPROVED** [review]
Thank you, @Toxicable!!!

**Review by IgorMinar — COMMENTED** [review]
dang! I just noticed that this PR also needs API docs update. I asked @kara to send a follow up PR so that we don't need to block this, but in the future @Toxicable please do update the API docs.

I also noticed that you wrote a great justification of this change in the issue, but that justification is not available in the commit message, in the future it would be easier for us to review these kinds of changes if the explanation was in the commit message. thank you

**Review by IgorMinar — COMMENTED** [review]
dang #2!

this commit should have been a feature or a fix since it changes the public api surface
otherwise the change won’t get highlighted in the changelog and nobody will know

**Comment by Toxicable** [comment]
@IgorMinar I can change it to a fix now if you like?

**Comment by kara** [comment]
@Toxicable If you can change it to a fix and add the justification to the commit message, that'd be awesome!

**Comment by Toxicable** [comment]
@kara done!

**Pull request closed without merging**